### PR TITLE
const interning: decide about mutability purely based on the kind of interning, not the types we see

### PR DESCRIPTION
**Pull request by RalfJung**
This entirely replaces our const-eval interner, i.e. the code that takes the final result of a constant evaluation from the local memory of the const-eval machine to the global `tcx` memory. The main goal of this change is to ensure that we can detect mutable references that sneak into this final value -- this is something we want to reject for `static` and `const`, and while const-checking performs some static analysis to ensure this, I would be much more comfortable stabilizing const_mut_refs if we had a dynamic check that sanitizes the final value. (This is generally the approach we have been using on const-eval: do a static check to give nice errors upfront, and then do a dynamic check to be really sure that the properties we need for soundness, actually hold.)

The new interner is a lot simpler than the old one: it recursively traverses the allocation holding the final result, and all allocations reachable from it, and ensures they all get interned. The initial allocation is interned as immutable for `const` and pomoted and non-interior-mutable `static`; all other allocations are interned as immutable for `static`, `const`, and promoted. The main subtlety is justifying that those inner allocations may indeed be interned immutably, i.e., that mutating them later would anyway already be UB:
- for promoteds, we rely on the analysis that does promotion to ensure that this is sound
- for `const` and `static`, we check that all pointers in the final result that point to things that are new (i.e., part of this const evaluation) are immutable, i.e., were created via `&<expr>` at a non-interior-mutable type.

Interning raises an error if it encounters a dangling pointer or a mutable pointer that violates the above rules.

I also extended our type-driven const validity checks to ensure that `&mut T` in the final value of a const points to mutable memory, at least if `T` is not zero-sized. This catches cases of people turning `&i32` into `&mut i32` (which would still be considered a read-only pointer). Similarly, when these checks encounter an `UnsafeCell`, they are checking that it lives in mutable memory. (Both of these only traverse the newly created values; if those point to other consts/promoteds, the check stops there. But that's okay, we don't have to catch all the UB.) I co-developed this with the stricter interner changes but I can split it out into a separate PR if you prefer.

This PR does have the immediate effect of allowing some new code on stable, for instance:
```rust
const CONST_RAW: *const Vec<i32> = &Vec::new() as *const _;
```
Previously that code got rejected since the type-based interner didn't know what to do with that pointer. It's a raw pointer, we cannot trust its type. The new interner does not care about types so it sees no issue with this code; there's an immutable pointer pointing to some read-only memory (storing a `Vec<i32>`), all is good. I can't think of a way to reject this code without type-based interning. Accepting this code pretty much commits us to non-type-based interning, but I think that's the better strategy anyway.

This PR also leads to slightly worse error messages when the final value of a const contains a dangling reference. Previously we would complete interning and then the type-based validation would detect this dangling reference and show a nice error saying where in the value (i.e., in which field) the dangling reference is located. However, the new interner cannot distinguish dangling references from dangling raw pointers, so it must throw an error when it encounters either of them. It doesn't have an understanding of the value structure so all it can say is "somewhere in this constant there's a dangling pointer". (Later parts of the compiler don't like dangling pointers/references so we have to reject them either during interning or during validation.) This could potentially be improved by doing validation before interning, but that's a larger change that I have not attempted yet. (It's also subtle since we do want validation to use the final mutability bits of all involved allocations, and currently it is interning that marks a bunch of allocations as immutable -- that would have to still happen before validation.)

### CODE REVIEW

**Comment by oli-obk** [comment]
> 1. This affects for instance `static mut FOO: &i32 = &0;`.

That uses promotion, but `static mut FOO: &i32 = {let x = 0; &{x}};` or something like it should do it

**Comment by oli-obk** [comment]
> 2\. and same for [this one](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0dac8d173a2b3099b9c2854fdad7a87c).

I think you meant https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=20d43a96a5f0b49e77da0a54a397633b which is rejected during interning due to a dangling pointer

**Comment by RalfJung** [comment]
> That uses promotion, but static mut FOO: &i32 = {let x = 0; &{x}}; or something like it should do it

I thought leading & is not promotion but the "outer scope rule"?

**Comment by oli-obk** [comment]
> I thought leading & is not promotion but the "outer scope rule"?

Promotion happens first 😅

**Comment by RalfJung** [comment]
Oh okay, so if promotion does not kick in then it falls back to outer scope, but if both apply promotion wins.

**Comment by RalfJung** [comment]
Okay, I added the test, as a Miri-unleash test to ensure this works even if the static checks are sleeping. The new checks are basically just slightly generalizing existing checks we already had to prevent UnsafeCell and &mut in constants.

I do wonder if there is any way one could have previously declared a static that contains an UnsafeCell, which would now be rejected. Might be worth a crater run.
@bors try

**Comment by RalfJung** [comment]
@craterbot check

**Review by oli-obk — APPROVED** [review]

**Comment by RalfJung** [comment]
I expect we'll see perf improvements when the type-based traversal is completely removed from the interner.

File: compiler/rustc_const_eval/src/interpret/validity.rs [hunk 1]
```diff
@@ -418,6 +438,21 @@ impl<'rt, 'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> ValidityVisitor<'rt, 'mir, '
...
+                PointerKind::Box => Mutability::Mut,
+                PointerKind::Ref => {
+                    let tam = value.layout.ty.builtin_deref(false).unwrap();
+                    // ZST never require mutability. We do not take into account interior mutability here
+                    // since we cannot know if there really is an `UnsafeCell` -- so we check that
+                    // in the recursive descent behind this reference.
+                    if size == Size::ZERO || tam.mutbl == Mutability::Not {
+                        Mutability::Not
+                    } else {
+                        Mutability::Mut
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/interpret/validity.rs:452]: We could have an extra check for `matches!(self.ctfe_mode, Some(Const(..)))` which then rejects mutable references immediately... but we already intern all allocations as immutable and so such code will already error, just with a different error message ("mutable ref in const" vs "dangling ref"/"mutable ref to immutable memory").

File: tests/ui/consts/miri_unleashed/mutable_references_err.64bit.stderr [hunk 2]
```diff
@@ -2,53 +2,79 @@ error[E0080]: it is undefined behavior to use this value
...
+   = note: the raw bytes of the constant (size: 8, align: 8) {
+               ╾───────alloc1────────╼                         │ ╾──────╼
+           }
+
+error[E0080]: it is undefined behavior to use this value
+  --> $DIR/mutable_references_err.rs:35:1
    |
 LL | const BLUNT: &mut i32 = &mut 42;
-   | ^^^^^^^^^^^^^^^^^^^^^ constructing invalid value: encountered mutable reference in a `const`
+   | ^^^^^^^^^^^^^^^^^^^^^ constructing invalid value: encountered mutable reference or box pointing to read-only memory
```
Comment (RalfJung) [inline tests/ui/consts/miri_unleashed/mutable_references_err.64bit.stderr:38]: Here, this is the potentially confusing error: we're creating an `&mut 42` and then we're told  it is pointing to read-only memory.

Should I bring back the "encountered mutable reference in `const"? And then, similarly, "encountered mutable reference in `static`"?

Comment (oli-obk) [reply tests/ui/consts/miri_unleashed/mutable_references_err.64bit.stderr:38]: Could we just add a note that says `all memory in a constant is immutable` or `all memory in a static must be immutable or use interior mutability`

...which reminds me: should we intern all nested allocs as mutable if a static's type is `!Freeze`? Just like with `static mut`

**Comment by RalfJung** [comment]
> ...which reminds me: should we intern all nested allocs as mutable if a static's type is !Freeze? Just like with static mut

I think no, whether inner allocs have interior mut vs whether the outermost alloc has interior mut is not correlated.

File: tests/ui/consts/miri_unleashed/mutable_references_err.64bit.stderr (hunk 2 above)
Comment (RalfJung) [reply tests/ui/consts/miri_unleashed/mutable_references_err.64bit.stderr:38]: > Could we just add a note that says all memory in a constant is immutable or all memory in a static must be immutable or use interior mutability

We could, but at that point why not do the more direct check?

Comment (RalfJung) [reply tests/ui/consts/miri_unleashed/mutable_references_err.64bit.stderr]: I did that now. The other check is also still needed, we have an unleash test for that.

File: tests/ui/consts/const-points-to-static.rs
Comment (RalfJung) [inline tests/ui/consts/const-points-to-static.rs:1]: I moved this into tests/ui/consts/const-points-to-static.rs, not sure why it was separate

**Comment by RalfJung** [comment]
I'll re-queue crater with the latest version.
@craterbot abort

**Comment by RalfJung** [comment]
@craterbot check

**Comment by RalfJung** [comment]
I was fooled by the tail expression rule (again! that rule is such a never-ending source of problems). The following gets accepted with miri-unleashed:
```rust
struct SyncPtr<T> { x : *const T }
unsafe impl<T> Sync for SyncPtr<T> {}

static RAW_SYNC: SyncPtr<AtomicI32> = SyncPtr { x: &AtomicI32::new(42) };
static RAW_MUT_CAST: SyncPtr<i32> = SyncPtr { x : &mut 42 as *mut _ as *const _ };
static RAW_MUT_COERCE: SyncPtr<i32> = SyncPtr { x: &mut 0 };
```
and after this PR the inner allocations all become read-only, which is probably not what the user expected. Crater won't be able to find this.

**Comment by RalfJung** [comment]
@craterbot run mode=check-only crates=https://crater-reports.s3.amazonaws.com/pr-116745/retry-regressed-list.txt

**Comment by RalfJung** [comment]
@craterbot p=1

**Comment by RalfJung** [comment]
Those are actually both legit regressions (and they look to be basically the same code so I only checked one).

https://github.com/phip1611/simple-chunk-allocator is using `feature(const_mut_refs)` so we are allowed to break this. It exploits that currently we reject `&mut` in the final value of `const`, but not `static`. I am not sure if we deliberately only did that check in const? This PR extends it to `static` as well.

So, from a breaking-change perspective this seems entirely fine. I still need to figure out what to do with [this](https://github.com/rust-lang/rust/pull/116745#issuecomment-1766070604) though.

**Comment by RalfJung** [comment]
Yeah, even bootstraping already fails due to [this](https://github.com/zbraniecki/pluralrules/blob/d0715d5b7cca72e228d6e6c4de154bd30dfdd06a/intl_pluralrules/src/rules.rs#L16) const. We can't take back the inner scope rule, the best we can do is nerf it around raw pointers. If in static/const we only use it for shared references without `UnsafeCell` then we should have sufficient justification for making everything immutable...

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs [hunk 3]
```diff
@@ -301,6 +301,9 @@ pub fn eval_in_interpreter<'mir, 'tcx>(
     cid: GlobalId<'tcx>,
     is_static: bool,
 ) -> ::rustc_middle::mir::interpret::EvalToAllocationRawResult<'tcx> {
+    // `is_static` just means "in static", it could still be a promoted!
+    debug_assert_eq!(is_static, ecx.tcx.static_mutability(cid.instance.def_id()).is_some());
```
Comment (RalfJung) [inline compiler/rustc_const_eval/src/const_eval/eval_queries.rs:305]: @oli-obk the `is_static` parameter here seems redundant, I couldn't quite figure out why we have it?

**Comment by RalfJung** [comment]
So... not sure what to do here.

Currently all cases I know of that are problematic here are rejected by the const checks: if the local being borrowed lives on beyond the end of this initializer, we reject interior mutability and mutable references even if the corresponding feature is enabled. Therefore, when we intern and we find allocations that still live and that haven't been interned yet, they must be from a shared reference without interior mutability, and hence we can intern them as immutable.

This seems like a pretty solid argument. However, unlike the rest of our const-eval story, we do not have a safety net here: the final checks we run on the value cannot determine if the memory was meant to be mutable or not.

The static checks are based on whether the local has `StorageDead`. Using custom-mir (if we ever support this for static/const bodies) one can easily have locals that do have `StorageDead` in dead code only, and then one can have a mutable reference survive until the end of the const/static initializer and have it interned immutably. Interning mutably is not an option for `const`.

I think the only way we can have sound runtime checks is to actually implement at least a crude aliasing model, which tracks whether a pointer is derived from a read-only reference, and allows only those kinds of pointers to be interned in the final constant.

---

And of course the other alternative is a type-driven interner. It doesn't need all the fancy machinery of our current interner I think, now that we know that nested inner allocations are only supposed to occur as shared references to non-interior-mutable data (plus the mess that is `static mut`). Maybe that is indeed the best option, since it rejects most programs...

@rustbot author

**Comment by RalfJung** [comment]
Ah no, even the type-driven interner is not enough with custom MIR. We could have something like
```rust
let mut local = (0, 0); // lives until the end
(&local.0, &mut local.1 as *mut _)
```
This would then intern `local` immutably since a shared reference points to it, but that's unsound since there's also a mutable raw pointer pointing there.

File: compiler/rustc_const_eval/src/const_eval/eval_queries.rs (hunk 3 above)
Comment (oli-obk) [reply compiler/rustc_const_eval/src/const_eval/eval_queries.rs:305]: 😆 just to avoid a second `is_static` invocation.

**Comment by RalfJung** [comment]
I think I'll try to implement a very lightweight aliasing model for CTFE -- tracking when a pointer was created as a shared reference, and is hence immutable. Then we can refuse to intern consts that have mutable uninterned pointers in their final value.

**Comment by RalfJung** [comment]
Marking as blocked on https://github.com/rust-lang/rust/pull/118324.

**Comment by RalfJung** [comment]
As I was hoping, getting rid of that type-driven traversal does help quite a bit with perf. :)

I'll close this PR though, it contains too much old discussion. Will start a new one.

**Pull request closed without merging**